api/src/test/java/org/openmrs/api/context/DaemonTest.java

```diff
@@ -52,6 +58,36 @@ public void executeScheduledTask_shouldNotThrowErrorIfCalledFromATimerSchedulerT
 		Assert.assertTrue(new PrivateSchedulerTask(task).runTheTest());
 	}
 	
+	@Test 
+	public void createUser_shouldThrowWhenCalledOutsideContextDAO() throws Throwable {
+		try {
+			Daemon.createUser(new User(), "password", null);
+			Assert.fail("Should not be here, an exception should have been thrown in the line above");
+		}
+		catch (APIException e) {
+			Assert.assertTrue(e.getMessage().startsWith(Context.getMessageSourceService().getMessage("Context.DAO.only", new Object[] { this.getClass().getName() }, null)));
+		}
+	}
+	
+	@Test
+	public void createUser_shouldCreateUserWithRolesInContextDAO() throws Throwable {
+		// setup
+		User u = new User();
+		u.setPerson(new Person());
+		u.addName(new PersonName("Jane", "X", "Doe"));
+		u.setUsername("jdoe");
+		u.getPerson().setGender("F");
+		
+		// replay
+		User createdUser = Context.getContextDAO().createUser(u, "P@ssw0rd", Arrays.asList("Provider", "Foobar Role"));
```

Comment:
@dkayiwa this is basically all that is needed. A way to create a new user to which a list of (existing) roles is added. This is exactly what a custom auth. scheme might do.

What is provided is a list of role names.
* If a role name does indeed correspond to an existing role in OpenMRS, then the role is added and the user is created with that role as part of its set of roles.
* If a role name is unknown to OpenMRS, that role is skipped and the user is created anyway.

So if all the roles names are bogus, or if no role names are provided, it is possible to end up with a user with no roles at all. I think that's fine, it's up to the implementers to ensure that they reference role names that make sense.